File: app/portfolio/portfolio_manager.py
from collections import defaultdict
from app.core.logger import logger
from app.cache.redis import redis_client


class PortfolioManager:

    positions = defaultdict(dict)

    @classmethod
    async def get_cash_balance(cls) -> float:
        val = await redis_client.get("portfolio:cash_balance")
        return float(val) if val else 100000.0

    @classmethod
    async def set_cash_balance(cls, amount: float):
        await redis_client.set("portfolio:cash_balance", str(amount))

    @classmethod
    async def get_realized_pnl(cls) -> float:
        val = await redis_client.get("portfolio:realized_pnl")
        return float(val) if val else 0.0

    @classmethod
    async def set_realized_pnl(cls, amount: float):
        await redis_client.set("portfolio:realized_pnl", str(amount))

    @classmethod
    async def load_positions(
        cls,
        positions_list: list
    ):
        cls.positions.clear()
        for pos in positions_list:
            cls.positions[pos.symbol] = {
                "quantity": pos.quantity,
                "average_price": pos.average_price,
                "stop_loss": pos.stop_loss,
                "take_profit": pos.take_profit
            }

        logger.info(
            f"Loaded {len(positions_list)} positions "
            f"into memory."
        )

    @classmethod
    async def update_position(
        cls,
        symbol: str,
        side: str,
        quantity: float,
        price: float
    ):
        current_position = cls.positions.get(symbol)
        cash = await cls.get_cash_balance()
        pnl = await cls.get_realized_pnl()

        if side == "BUY":
            cash -= (quantity * price)
            cls.positions[symbol] = {
                "quantity": quantity,
                "average_price": price,
                "stop_loss": price * 0.98,
                "take_profit": price * 1.03
            }
            await cls.set_cash_balance(cash)

        elif side == "SELL" and current_position:
            trade_pnl = (price - current_position["average_price"]) * quantity
            pnl += trade_pnl
            cash += (quantity * price)
            
            cls.positions.pop(symbol)
            await cls.set_realized_pnl(pnl)
            await cls.set_cash_balance(cash)

            logger.info(f"Trade Closed. PnL={trade_pnl}")

        logger.info(f"Portfolio Updated: Cash={cash}")

File: backend/src/trading/portfolio/portfolioManager.ts
import redisClient from "../../core/redis";
import { logger } from "../../core/logger";

interface PositionState {
  quantity: number;
  average_price: number;
  stop_loss?: number | null;
  take_profit?: number | null;
}

export class PortfolioManager {
  static positions: { [symbol: string]: PositionState } = {};

  static async getCashBalance(): Promise<number> {
    const val = await redisClient.get("portfolio:cash_balance");
    return val ? parseFloat(val) : 100000.0;
  }

  static async setCashBalance(amount: number): Promise<void> {
    await redisClient.set("portfolio:cash_balance", String(amount));
  }

  static async getRealizedPnl(): Promise<number> {
    const val = await redisClient.get("portfolio:realized_pnl");
    return val ? parseFloat(val) : 0.0;
  }

  static async setRealizedPnl(amount: number): Promise<void> {
    await redisClient.set("portfolio:realized_pnl", String(amount));
  }

  static async loadPositions(positionsList: any[]): Promise<void> {
    this.positions = {};
    for (const pos of positionsList) {
      this.positions[pos.symbol] = {
        quantity: pos.quantity,
        average_price: pos.average_price,
        stop_loss: pos.stop_loss,
        take_profit: pos.take_profit,
      };
    }
    logger.info(`Loaded ${positionsList.length} positions into memory.`);
  }

  static async updatePosition(
    symbol: string,
    side: string,
    quantity: number,
    price: number
  ): Promise<void> {
    const currentPosition = this.positions[symbol];
    let cash = await this.getCashBalance();
    let pnl = await this.getRealizedPnl();

    if (side === "BUY") {
      cash -= (quantity * price);
      this.positions[symbol] = {
        quantity,
        average_price: price,
        stop_loss: price * 0.98,
        take_profit: price * 1.03,
      };
      await this.setCashBalance(cash);
    } else if (side === "SELL" && currentPosition) {
      const tradePnl = (price - currentPosition.average_price) * quantity;
      pnl += tradePnl;
      cash += (quantity * price);

      delete this.positions[symbol];
      await this.setRealizedPnl(pnl);
      await this.setCashBalance(cash);

      logger.info(`Trade Closed. PnL=${tradePnl}`);
    }

    logger.info(`Portfolio Updated: Cash=${cash}`);
  }
}
export default PortfolioManager;
